lib/Twig/Cache/Filesystem.php

```diff
@@ -36,9 +36,8 @@ public function __construct($directory, $options = 0)
      */
     public function generateKey($name, $className)
     {
-        $hash = hash('sha256', $className);
-
-        return $this->directory.'/'.$hash[0].'/'.$hash[1].'/'.$hash.'.php';
+        // We assume the class name is suffixed by a hash (see Twig_Environment::getTemplateClass) that we can use to balance the directory tree.
+        return $this->directory.'/'.substr($className, -2).'/'.$className.'.php';
```

Comment:
I know. I'm just trying to make it as fast as possible by default and avoid hashing the hash for every template on each request. Maybe we should remove the directory balancing at all by default. Todays filesystems do not have a limitation on number of files per directory. And also most users, won't have thousands of templates anyway where it could matter. And if people really have this need they can change the filesystem cache themselves. No need to hurt everybody by default.